lib/plugins/aws/lib/naming.js

```diff
@@ -0,0 +1,234 @@
+ * MyFuncLambdaFunctionArn)  Generally, normalization of the name also occurs.
+ *
+ * Sometimes, it is important to deconstruct the names that are created via this centralized
+ * utility.  As a result, the regular expressions used for identifying specific resource types using
+ * their names or the logic to extract a name components from a logical name are things that must
+ * also be centralized so that they co-occur and are more correlated spatially within the code base.
+ * That is to say that they are easier to change together.
+ */
+class Naming {
+  // TODO this configure pattern is fugly.  it serves the need to have static and also configured
+  // TODO use patterns without complicating the life of consumers.  In standard (non-test) usage
+  // TODO the frequency of using this method should be low so maybe it doesn't matter.
+  // TODO better suggestions highly welcome
+  configure(serverless) {
+    this.serverless = serverless;
+  }
+
+  /*
+   * General Name Normalization
+   */
+  normalizeName(name) {
+    return `${_.upperFirst(name)}`;
+  }
+  normalizeNameToAlphaNumericOnly(name) {
+    return this.normalizeName(name.replace(/[^0-9A-Za-z]/g, ''));
+  }
+  normalizePathtoCapitalAlphaNumbericOnlyWithReplacement(path) {
+    return this.normalizeNameToAlphaNumericOnly(
+      path.replace(/-/g, 'Dash')
+          .replace(/\{(.*)\}/g, '$1Var'));
+  }
+
+  // Probably belongs in a non-aws-specific code asset
+  getServiceEndpointRegex() {
+    return /^ServiceEndpoint/;
+  }
+
+  /*
+   * Stack Naming
+   */
+  getStackName() {
+    return `${this.serverless.service.service}-${this.serverless.config.stage}`;
+  }
+
+  /*
+   * IAM (Role and Policy) Naming
+   */
+  getRolePath() {
+    return '/';
+  }
+  getRoleName() {
+    return {
+      'Fn::Join': [
+        '-',
+        [
+          this.serverless.service.service,
+          this.serverless.config.stage,
+          this.serverless.config.region,
+          'lambdaRole',
+        ],
+      ],
+    };
+  }
+  getLogicalRoleName() {
+    return 'IamRoleLambdaExecution';
+  }
+  getPolicyName() {
+    return { // TODO should probably have name ordered and altered as above - see AWS docs
+      'Fn::Join': [
+        '-',
+        [
+          this.serverless.config.stage,
+          // TODO is it a bug here in that the role is not specific to the region?
+          this.serverless.service.service,
+          'lambda',
+        ],
+      ],
+    };
+  }
+  getLogicalPolicyName() {
+    return 'IamPolicyLambdaExecution';
+  }
+
+  getLogGroupName(functionName) {
+    return `/aws/lambda/${functionName}`;
+  }
+
+  /*
+   * Lambda Function Naming
+   */
+  getNormalizedLambdaName(functionName) {
+    return this.normalizeName(functionName);
+  }
+  extractLambdaNameFromArn(functionArn) {
+    const splitArn = functionArn.split(':');
+    // TODO the following two lines assumes default function naming?  Is there a better way?
```

Comment:
Sorry, I think I wasn't quite clear.  This is from [authorizers.js#30](https://github.com/serverless/serverless/blob/master/lib/plugins/aws/deploy/compile/events/apiGateway/lib/authorizers.js#L30).  It is extracting a function name from a given ARN that should identify a lambda that is to be used as the authorizer for apig.  I don't believe that should contain the logical name that was used in the CFT to create the Lambda.  Instead, I believe that if I set a function's custom `name` to be something like `a-name-with-dashes` that I'll end up with an ARN that is something like `arn:aws:lambda:region:${accountId}:function:a-name-with-dashes`.

I would expect this code to return `dashes` rather than `a-name-with-dashes` which I believe is the expected extraction.

Another thought is that the following would seem to be more a appropriate authorizer ARN specification (just adding stage):
`arn:aws:lambda:region:${accountId}:function:a-name-with-dashes:${stage}`
In that case, the stage would be what gets split.

Sorry for being a bit confused and context-poor here, just asking questions from my curiosity while trying to make sure and capture all the name provision and manipulation logic.